In [33]:
import numpy as np
import astropy.units as u
from ReadFile import Read
from tabulate import tabulate
import latextable

In [11]:
def ComponentMass(filename, particle_type):
    """ This function returns the total mass of any galaxy component 
        from a given txt file with mass of each particle in the galaxy.
        
        Inputs: 
            filename: 'string'
                name of text file from which data need to be extracted
            
            particle_type: 'float'
                a float value specifies the type of the particle
                1.0 = dark matter
                2.0 = disk star
                3.0 = bulge star
        
        Output:
            totalmass: 'astropy quantity'
                total mass of the galaxy component in units of 
                10e12 Msun, rounded to 3 decimal places                
    """
    # extract the data array from a predefined function
    data = (Read(filename)[2])
    
    # create new indexing based on the particle type 
    if (particle_type == 1.0): 
        index = np.where(data["type"]==1.0)
    if (particle_type == 2.0):    
        index = np.where(data["type"]==2.0)
    if (particle_type == 3.0):    
        index = np.where(data["type"]==3.0)
        
    mass_array = data['m'][index] # define new arrays with mass of only
                            # required partcile type
    
    # Add mass of all particles of a given type to get the total mass
    mass = 0 
    for particle_mass in mass_array:
        mass = mass + particle_mass
    
    # mass is given in units of 1e10 Msun
    totalmass = mass*1e10*u.M_sun.to(1e12*u.Msun) # add required units 
    totalmass = np.around(totalmass,3) # round to 3 decimal places
    
    return totalmass

In [16]:
# Mass of each component for MW galaxy

MW_Halo = ComponentMass("MW_000.txt",1.0)
MW_Disk = ComponentMass("MW_000.txt",2.0)
MW_Bulge = ComponentMass("MW_000.txt",3.0)
MW_Total = MW_Halo+MW_Disk+MW_Bulge

# baryon fraction for MW
MW_f = np.round(((MW_Disk+MW_Bulge)/MW_Total), 3) 

In [17]:
# Mass of each component for M31 galaxy

M31_Halo = ComponentMass("M31_000.txt",1.0)
M31_Disk = ComponentMass("M31_000.txt",2.0)
M31_Bulge = ComponentMass("M31_000.txt",3.0)
M31_Total = M31_Halo+M31_Disk+M31_Bulge

# baryon fraction for M31
M31_f = np.round(((M31_Disk+M31_Bulge)/M31_Total), 3) 

In [18]:
# Mass of each component for M33 galaxy

M33_Halo = ComponentMass("M33_000.txt",1.0)
M33_Disk = ComponentMass("M33_000.txt",2.0)
M33_Bulge = ComponentMass("M33_000.txt",3.0)
M33_Total = M33_Halo+M33_Disk+M33_Bulge

M33_f = np.round(((M33_Disk+M33_Bulge)/M33_Total),3) # baryon fraction for M33

In [19]:
# Mass for local Group

Local_Halo = MW_Halo + M31_Halo +M33_Halo
Local_Disk = MW_Disk +M31_Disk + M33_Disk
Local_Bulge = MW_Bulge +M31_Bulge + M33_Bulge
Local_Total = Local_Halo + Local_Disk + Local_Bulge

# baryon fraction Local Group
Local_f = np.round(((Local_Disk+Local_Bulge)/Local_Total), 3) 

In [43]:
# Column names for table
Column_names = ["Galaxy name", "Halo mass \n [10^{12} M_sun]", 
                'Disk Mass \n [10^{12} M_sun]', 
                "Bulge Mass \n [10^{12} M_sun]", 
               "Total Mass \n [10^{12} M_sun]", "f_(bar)"]

# Data Array for table
Data = [["Milky Way",MW_Halo,MW_Disk, MW_Bulge, MW_Total, MW_f], 
        ["M31",M31_Halo,M31_Disk, M31_Bulge, M31_Total, M31_f], 
        ["M33",M33_Halo,M33_Disk, M33_Bulge, M33_Total, M33_f],
        ["Local Group", Local_Halo, Local_Disk, Local_Bulge, Local_Total, Local_f]]

# create table
'''
# latex for table
print('\nTabulate Latex:')
print(tabulate(Data, headers=Column_names, tablefmt='latex'))
'''
print(tabulate(Data, headers=Column_names, tablefmt="fancy_grid"))

╒═══════════════╤════════════════════╤════════════════════╤════════════════════╤════════════════════╤═══════════╕
│ Galaxy name   │         Halo mass  │         Disk Mass  │        Bulge Mass  │        Total Mass  │   f_(bar) │
│               │    [10^{12} M_sun] │    [10^{12} M_sun] │    [10^{12} M_sun] │    [10^{12} M_sun] │           │
╞═══════════════╪════════════════════╪════════════════════╪════════════════════╪════════════════════╪═══════════╡
│ Milky Way     │              1.975 │              0.075 │              0.01  │              2.06  │     0.041 │
├───────────────┼────────────────────┼────────────────────┼────────────────────┼────────────────────┼───────────┤
│ M31           │              1.921 │              0.12  │              0.019 │              2.06  │     0.067 │
├───────────────┼────────────────────┼────────────────────┼────────────────────┼────────────────────┼───────────┤
│ M33           │              0.187 │              0.009 │              0     │        

## Questions 4

#### 1]
Total Mass of both Milky Way and M31 is $2.06*10^{12} M_\odot$. They both have approximately same total mass and for both galaxies, the maojority of the mass is dominated y dark matter halo. 

#### 2]
Milky Way has a stellar mass of $0.085*10^{12} M_\odot$ while M31 has stellar mass of $0.139*10^{12} M_\odot$. M31 has slightly higher stellar mass than Milky Way. Since M31 has more stellar mass it should be comparitively more luminuous than Milky Way.

#### 3]

In [25]:
HaloRatio_MW_M31 = MW_Halo/M31_Halo
print(HaloRatio_MW_M31)

1.028110359187923


Milky Way has slightly larger fraction of Dark Matter Mass than M31 which is expected as for the same total mass, M31 has a slightly larger stellar mass.

#### 4]
From the table we can se that for each galaxy the baryonic fraction $f_{bar}$ is,

Milky Way : 0.041 

M31 : 0.067       
M33 : 0.046       
Local Group : 0.054

For all the galaxies this ratio is approximately (4-6)%. However, in universe ~16% of fraction is baryonic matter. This is 4 times than what we computed to be in a galaxy.
There is smmaler fraction of dark matter compared to baryonic matter in universe than in galaxies. This could be beacuse, in galaxies the dark matter halo are densely bounded to the galaxy but in larger scales of universe, the ratio does not hold. 